cloudinit/url_helper.py

```diff
@@ -382,6 +385,74 @@ def _cb(url):
     return None  # Should throw before this...
 
 
+def dual_stack(
+    func: Callable[..., Any],
+    addresses: List[str],
+    stagger_delay: float = 0.150,
+    timeout: int = 10,
+) -> Tuple:
+    """execute multiple callbacks in parallel
+
+    Run blocking func against two different addresses staggered with a
+    delay. The first call to return is returned from this function and
+    remaining unfinished calls are canceled (or finish asynchronously on
+    python <3.9
+    """
+    return_result = None
+    returned_address = None
+
+    def _run_func(func, addr, timeout, delay=None):
+        """Execute func with optional delay"""
+        if delay:
+            time.sleep(delay)
+        return func(addr, timeout)
+
+    executor = ThreadPoolExecutor(max_workers=len(addresses))
+    try:
+        futures = {
+            executor.submit(
+                _run_func,
+                func=func,
+                addr=addr,
+                timeout=timeout,
+                delay=(i * stagger_delay),
+            ): addr
+            for i, addr in enumerate(addresses)
+        }
+
+        # handle the first function to complete from the threadpool executor
+        future = next(as_completed(futures, timeout=timeout))
+
+        returned_address = futures[future]
+        return_result = future.result()
+        return_exception = future.exception()
+        if return_exception:
+            LOG.warning(
+                "Exception %s during request to %s",
+                return_exception,
+                returned_address,
+            )
+            raise return_exception
```

Comment:
Would we want to specifically look at requests.ConnectionError failures which basically tell us no route to host for IPv6 and ignore them as no route means we really aren't ever going to succeed there right?

Reply:
Interesting question. I don't think that "no route to host" should ever happen (because we're using link-local addresses[1]), but maybe I'm wrong. Generally the idea of "ignore connection failures in case the others succeed" is something I can get behind.

[1] To my knowledge Linux doesn't require fib entries for sending directly to a link-local addresses.

A quick experiment:
```
[ec2-user@ip-172-31-6-134 ~]$ ip r
default via 172.31.0.1 dev eth0 
169.254.169.254 dev eth0 
172.31.0.0/20 dev eth0 proto kernel scope link src 172.31.6.134 
[ec2-user@ip-172-31-6-134 ~]$ sudo ip r del 169.254.169.254
[ec2-user@ip-172-31-6-134 ~]$ ip route show table all | grep -e 169 -e ec2
[ec2-user@ip-172-31-6-134 ~]$ wget http://169.254.169.254
--2022-01-25 23:39:54--  http://169.254.169.254/
Connecting to 169.254.169.254:80... connected.
HTTP request sent, awaiting response... 200 OK
```
Even though the link local route was removed from the routing table it was still able to communicate using that address.

Reply:
Ahh maybe I was misinterpreting the log I saw on an IPv4 only instance:
`2022-01-25 22:34:46,195 - url_helper.py[WARNING]: Calling 'None' failed [0/120s]: request error [HTTPConnectionPool(host='fd00:ec2::254', port=80): Max retries exceeded with url: /latest/api/token (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7facb9c98400>: Failed to establish a new connection: [Errno 101] Network is unreachable'))]`

Reply:
Thanks for the context. "Network is unreachable" is often related to routing issues, I see what you were getting at. Regardless of what "Network is unreachable" is caused by,  I think this suggestion is a good one. Specifically this would make connectivity possible in the case that there is a slow first address and an unreachable second address.

Reply:
> Interesting question. I don't think that "no route to host" should ever happen (because we're using link-local addresses[1]), but maybe I'm wrong. 

Update: I was wrong.

My former argument here was invalid. The link local address used on ec2 instances is actually the destination address, not source address. Http(s) traffic leaves via the default route on a private address (RFC1918).

Remove default routes and this error could happen:

```
[ec2-user@ip-172-31-6-134 ~]$ wget http://[fd00:ec2::254]/latest/meta-data/
--2022-02-10 22:36:17--  http://[fd00:ec2::254]/latest/meta-data/
Connecting to [fd00:ec2::254]:80... failed: Network is unreachable.
```

I still think that the route I deleted in the example above (`169.254.169.254 dev eth0`) is unneeded, but for different reasons.

This comment doesn't imply a different course of action, but I wanted to set the record straight here.